pkg/edgecache/finalizer/controller.go

```diff
@@ -0,0 +1,216 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/component-base/metrics/prometheus/ratelimiter"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/blob-csi-driver/pkg/edgecache"
+	"sigs.k8s.io/blob-csi-driver/pkg/util"
+)
+
+const (
+	finalizerName          string = "external/blob-csi-driver-edgecache"
+	pvcProtectionFinalizer string = "kubernetes.io/pvc-protection"
+	accountAnnotation      string = "external/edge-cache-account"
+	containerAnnotation    string = "external/edge-cache-container"
+)
+
+// Controller is controller that removes PVProtectionFinalizer
+// from PVs that are not bound to PVCs.
+type Controller struct {
+	client clientset.Interface
+
+	pvcLister       corelisters.PersistentVolumeClaimLister
+	pvcListerSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+
+	eventRecorder record.EventRecorder
+
+	ecManager edgecache.ManagerInterface
+}
+
+// NewHydraFinalizerController returns a new *Controller.
+func NewHydraFinalizerController(manager edgecache.ManagerInterface, pvcInformer coreinformers.PersistentVolumeClaimInformer, client clientset.Interface) *Controller {
+	c := &Controller{
+		client: client,
+		queue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "hydrafinalizer"),
+	}
+	if client == nil {
+		klog.Error("Hydra Finalizer requires client")
+		return nil
+	}
+	if manager == nil {
+		klog.Error("Hydra Finalizer requires edgecache manager")
+		return nil
+	}
+	limiter := client.CoreV1().RESTClient().GetRateLimiter()
+	if limiter != nil {
+		err := ratelimiter.RegisterMetricAndTrackRateLimiterUsage("hydra_finalizer_controller", limiter)
+		if err != nil {
+			klog.Errorf("Unable to register ratelimiter")
+			return nil
+		}
+	}
+
+	c.ecManager = manager
+	c.pvcLister = pvcInformer.Lister()
+	c.pvcListerSynced = pvcInformer.Informer().HasSynced
+
+	pvcInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.pvcUpdated,
+	})
+
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartLogging(klog.Infof)
+	broadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: client.CoreV1().Events(v1.NamespaceAll)})
+	c.eventRecorder = broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "hydrafinalizer"})
+
+	klog.V(2).Infof("NewHydraFinalizerController: complete")
+	return c
+}
+
+// Run runs the controller goroutines.
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting Hydra Finalizer controller")
```

Comment:
same comment on naming conventions in various log messages here and below.

Reply:
updated